File: app/api/repos/catalog.py
# app/api/repos/catalog.py

from __future__ import annotations

from typing import Any

from app.core.neo4j import Neo4jClient


class CatalogRepo:
    """
    Data access layer for catalog (autocomplete/dropdown) endpoints.
    Returns raw data from Neo4j.
    """

    def __init__(self, neo4j_client: Neo4jClient) -> None:
        self.neo4j = neo4j_client

    def search_occupations_autocomplete(
        self, q: str | None, limit: int
    ) -> list[dict[str, Any]]:
        """
        Search occupations by label for autocomplete.

        Returns:
            List of dicts with keys: uri, label
        """
        cypher = """
        MATCH (o:Occupation)
        WHERE o.preferredLabel IS NOT NULL
          AND ($q IS NULL OR toLower(o.preferredLabel) CONTAINS toLower($q))
        RETURN o.uri AS uri, o.preferredLabel AS label
        ORDER BY o.preferredLabel
        LIMIT $limit
        """
        return self.neo4j.run_query(cypher, {"q": q, "limit": limit})

    def search_skills_autocomplete(
        self, q: str | None, limit: int
    ) -> list[dict[str, Any]]:
        """
        Search skills by label for autocomplete.

        Returns:
            List of dicts with keys: uri, label
        """
        cypher = """
        MATCH (s:Skill)
        WHERE s.preferredLabel IS NOT NULL
          AND ($q IS NULL OR toLower(s.preferredLabel) CONTAINS toLower($q))
        RETURN s.uri AS uri, s.preferredLabel AS label
        ORDER BY s.preferredLabel
        LIMIT $limit
        """
        return self.neo4j.run_query(cypher, {"q": q, "limit": limit})

    def search_occupation_groups(
        self, q: str | None, limit: int
    ) -> list[dict[str, Any]]:
        """
        Search occupation groups (ISCO groups) by label.

        Returns:
            List of dicts with keys: uri, code, label
        """
        cypher = """
        MATCH (g:OccupationGroup)
        WHERE $q IS NULL OR toLower(g.label) CONTAINS toLower($q)
        RETURN g.uri AS uri, g.label AS label, g.code AS code
        ORDER BY toLower(g.label)
        LIMIT $limit
        """
        return self.neo4j.run_query(cypher, {"q": q, "limit": limit})

    def search_skill_groups(
        self, q: str | None, limit: int
    ) -> list[dict[str, Any]]:
        """
        Search skill groups by label.

        Returns:
            List of dicts with keys: uri, label
        """
        cypher = """
        MATCH (g:SkillGroup)
        WHERE $q IS NULL OR toLower(g.label) CONTAINS toLower($q)
        RETURN g.uri AS uri, g.label AS label
        ORDER BY toLower(g.label)
        LIMIT $limit
        """
        return self.neo4j.run_query(cypher, {"q": q, "limit": limit})

    def list_concept_schemes(self) -> list[dict[str, Any]]:
        """
        List all concept schemes.

        Returns:
            List of dicts with keys: uri, label
        """
        cypher = """
        MATCH (cs:ConceptScheme)
        RETURN cs.uri AS uri, cs.label AS label
        ORDER BY toLower(cs.label)
        """
        return self.neo4j.run_query(cypher)

File: app/frontend/services/catalog.ts
import api from '@/lib/api';
import type { AutocompleteOption } from '@/types';

export interface OccupationGroup {
    uri: string;
    code: string;
    label: string;
}

export interface SkillGroup {
    uri: string;
    label: string;
}

export interface ConceptScheme {
    uri: string;
    label: string;
}

// Cache to store the full list of occupation groups
let occupationGroupsCache: OccupationGroup[] | null = null;
let isFetchingGroups = false;

export const catalogService = {
    searchSkills: async (q: string, limit = 10): Promise<AutocompleteOption[]> => {
        const response = await api.get<AutocompleteOption[]>('/catalog/skills', {
            params: { q, limit }
        });
        return response.data;
    },

    searchOccupations: async (q: string, limit = 10): Promise<AutocompleteOption[]> => {
        const response = await api.get<AutocompleteOption[]>('/catalog/occupations', {
            params: { q, limit }
        });
        return response.data;
    },

    // Get occupation label from URI 
    getOccupationLabel: async (occupationUri: string): Promise<string> => {
        try {
            const response = await api.get('/occupations', {
                params: { limit: 100 }
            });
            const occupations = response.data;
            const match = occupations.find((o: any) => o.uri === occupationUri);
            if (match) return match.label;

            const parts = occupationUri.split('/');
            const lastPart = parts[parts.length - 1];
            const searchResults = await catalogService.searchOccupations(lastPart, 50);
            const searchMatch = searchResults.find(o => o.uri === occupationUri);
            return searchMatch?.label || occupationUri;
        } catch {
            return occupationUri;
        }
    },

    getOccupationGroups: async (q?: string, limit = 100): Promise<OccupationGroup[]> => {
        // If searching, use the normal API
        if (q) {
            const response = await api.get<OccupationGroup[]>('/catalog/occupation-groups', {
                params: { q, limit }
            });
            return response.data;
        }

        // If not searching and we have a cache, return it
        if (occupationGroupsCache) {
            return occupationGroupsCache;
        }

        // If we're already fetching, wait for it (simple promise-based lock could be added here if needed)
        // For now, we'll just implement the aggregator
        if (isFetchingGroups) return [];

        try {
            isFetchingGroups = true;

            // Recursive crawler to fetch all groups despite backend limits
            const fetchRecursively = async (prefix: string): Promise<OccupationGroup[]> => {
                const limit = 100;
                const response = await api.get<OccupationGroup[]>('/catalog/occupation-groups', {
                    params: { q: prefix, limit }
                });
                const data = response.data;

                // If we hit the limit, we likely missed some items. Drill down.
                // We stop recursing at length 2 to avoid explosion (26*26 = 676 requests max worst case)
                if (data.length === limit && prefix.length < 2) {
                    const suffixes = 'abcdefghijklmnopqrstuvwxyz'.split('');
                    const subResults = await Promise.all(
                        suffixes.map(char => fetchRecursively(prefix + char))
                    );
                    return subResults.flat();
                }

                return data;
            };

            // Start with base alphanumeric characters
            // We search for single letters. If any letter returns 100 items, the recursive function will split it.
            const baseKeys = 'abcdefghijklmnopqrstuvwxyz0123456789'.split('');
            const results = await Promise.all(baseKeys.map(key => fetchRecursively(key)));

            const allGroups = results.flat();
            // Deduplicate by URI
            const uniqueGroups = Array.from(new Map(allGroups.map(item => [item.uri, item])).values());

            // Sort alphabetically
            uniqueGroups.sort((a, b) => a.label.localeCompare(b.label));

            occupationGroupsCache = uniqueGroups;
            return uniqueGroups;
        } finally {
            isFetchingGroups = false;
        }
    },

    getSkillGroups: async (q?: string, limit = 50): Promise<SkillGroup[]> => {
        const response = await api.get<SkillGroup[]>('/catalog/skill-groups', {
            params: { q, limit }
        });
        return response.data;
    },

    getConceptSchemes: async (): Promise<ConceptScheme[]> => {
        const response = await api.get<ConceptScheme[]>('/catalog/concept-schemes');
        return response.data;
    }
};
